### PR TITLE
Introduce memory-matcher based rules

### PR DESCRIPTION
**Pull request by michaelabuckley**

### CODE REVIEW

**Comment by tadgh** [comment]
You may want to check in with @JasonRoberts-smile who is also working on the InMemoryResourceMatcher. I imagine this change will bust his MR

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/docs/security/authorization_interceptor.md [hunk 1]
```diff
@@ -19,6 +19,15 @@ The AuthorizationInterceptor is used by subclassing it and then registering your
 {{snippet:classpath:/ca/uhn/hapi/fhir/docs/AuthorizationInterceptors.java|patientAndAdmin}}
 ``` 
 
+The core rules support restricting access by resource type, resource instance, and compartment.
+The rules also support fhir-query filters expressed by fhir queries - e.g. `code:above=http://loinc.org|55399-0` to restrict Observations to just the diabetes panel.
+To use the fhir-query filters, you must activate the [RuleFilteringConsentService](/apidocs/hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/consent/RuleFilteringConsentService.java)
+like the
```
Comment (michaelabuckley) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/docs/security/authorization_interceptor.md:25]:
```suggestion
The core rules support restricting access by resource type, resource instance, and compartment.
The rules also support query filters expressed by FHIR queries - e.g. `code:above=http://loinc.org|55399-0` to restrict Observations to just the diabetes panel.
To use thequery filters, you must activate the [RuleFilteringConsentService.](/apidocs/hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/consent/RuleFilteringConsentService.java)
```

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/AuthorizationInterceptor.java [hunk 2]
```diff
@@ -177,6 +179,16 @@ public AuthorizationInterceptor setValidationSupport(IValidationSupport theValid
...
 
+	/**
+	 * Sets a search parameter matcher for use in handling SMART v2 filter scopes
+	 *
+	 * @param myAuthorizationSearchParamMatcher The search parameter matcher. Null is also acceptable and is used as the
+	 *                                          default.   
+	 */
+	public void setMyAuthorizationSearchParamMatcher(IAuthorizationSearchParamMatcher myAuthorizationSearchParamMatcher) {
+		this.myAuthorizationSearchParamMatcher = myAuthorizationSearchParamMatcher;
+	}
```
Comment (michaelabuckley) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/AuthorizationInterceptor.java:190]:
```suggestion
	 * @param theAuthorizationSearchParamMatcher The search parameter matcher. Null is also acceptable and is used as the
	 *                                          default.   
	 */
	public void setMyAuthorizationSearchParamMatcher(IAuthorizationSearchParamMatcher theAuthorizationSearchParamMatcher) {
		this.myAuthorizationSearchParamMatcher = theAuthorizationSearchParamMatcher;
	}
```

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_1_0/3674-implement-rule-filters-and-consent.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+type: add
+issue: 3674
+title: "The rule system has been extended to support optional fhir-expressions than narrow the application a rule.
```
Comment (fil512) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/6_1_0/3674-implement-rule-filters-and-consent.yaml:4]:
```suggestion
title: "The rule system has been extended to support optional fhir-expressions that narrow the application of a rule.
```

File: hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/docs/security/authorization_interceptor.md [hunk 4]
```diff
@@ -19,6 +19,14 @@ The AuthorizationInterceptor is used by subclassing it and then registering your
 {{snippet:classpath:/ca/uhn/hapi/fhir/docs/AuthorizationInterceptors.java|patientAndAdmin}}
 ``` 
 
+The core rules support restricting access by resource type, resource instance, and compartment.
+The rules also support query filters expressed by FHIR queries - e.g. `code:above=http://loinc.org|55399-0` to restrict Observations to just the diabetes panel.
+To use thequery filters, you must activate the [RuleFilteringConsentService.](/apidocs/hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/consent/RuleFilteringConsentService.java)
```
Comment (fil512) [inline hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/docs/security/authorization_interceptor.md:24]:
```suggestion
To use query filters, you must activate the [RuleFilteringConsentService.](/apidocs/hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/consent/RuleFilteringConsentService.java)
```

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/AuthorizationInterceptorJpaR4Test.java [hunk 5]
```diff
@@ -1749,4 +1749,22 @@ public List<IAuthRule> buildRuleList(RequestDetails theRequestDetails) {
 			.resourceConditionalByUrl("Patient?name=Siobhan&_expunge=true")
 			.execute();
 	}
+
+	@Test
+	public void testSmartFilterSearch() {
+		// wipjv flesh out unit tests for new rule type here.
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/AuthorizationInterceptorJpaR4Test.java:1755]: wip

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/AuthorizationInterceptorJpaR4Test.java [hunk 6]
```diff
@@ -1749,4 +1749,22 @@ public List<IAuthRule> buildRuleList(RequestDetails theRequestDetails) {
...
+			public List<IAuthRule> buildRuleList(RequestDetails theRequestDetails) {
+				return new RuleBuilder()
+					.allow().read().allResources().inCompartmentWithFilter("Patient", new IdType("Patient/bob"), "date=gt2022-01-01").andThen()
+					.denyAll()
+					.build();
+			}
+		});
+
+	   // when
+		myClient.search().byUrl("/Observation?code=foo&patient=Patient/bob").returnBundle(Bundle.class).execute();
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/provider/r4/AuthorizationInterceptorJpaR4Test.java:1768]: missing assert

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/searchparam/SmartV2AuthServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+		public void basicMatch_unsupported() {
+			// given
+			helpCreatePatient();
+
+			// when
+			result = myMatcher.match("Patient?birthdate=ap2020", myResource);
+
+			// then
+			assertThat(result, notNullValue());
+			assertThat(result.getMatch(), equalTo(IAuthorizationSearchParamMatcher.Match.UNSUPPORTED));
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/searchparam/SmartV2AuthServiceTest.java:135]: nice

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/searchparam/SmartV2AuthServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+			// given
+			helpCreatePatient();
+			FhirQueryRuleImpl rule = buildSimplePatientRuleWithFilter("birthdate=ap2020");
+
+			// when
+			AuthorizationInterceptor.Verdict verdict = rule.applyRule(RestOperationTypeEnum.SEARCH_TYPE, mySrd, null, null, myResource, myRuleApplier, new HashSet<>(), Pointcut.STORAGE_PRESHOW_RESOURCES);
+
+			// then
+			assertThat(verdict, nullValue());
+			assertThat(myLogCapture.getLogEvents(),
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/searchparam/SmartV2AuthServiceTest.java:180]: nice.  feels like the right choice.

Comment (fil512) [reply hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/searchparam/SmartV2AuthServiceTest.java:180]: I just hope the warning message has enough context that the user will be able to track down the root cause

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImplTest.java [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.hasItem;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+import static org.mockito.Mockito.when;
+
+// wipjv where should this test live? -
+// wipjv can we mock the resource?  We just use it for stubbing here. If so, move this back to hapi-fhir-server ca.uhn.fhir.rest.server.interceptor.auth
+@MockitoSettings(strictness= Strictness.LENIENT)
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImplTest.java:34]: not a fan of lenient mocks.  Can you use @Spy instead?

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImplTest.java [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+	public class MatchingLogic {
+		@BeforeEach
+		public void setUp() {
+			withSearchParamMatcherPresent();
+		}
+
+		@Test
+		public void simpleStringSearch_match_allow() {
+			// given
+			withPatientWithNameAndId();
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImplTest.java:72]: Rename this method to withPatientWithSmithNameAndId(); so it's really clear this is the matching scenario

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImplTest.java [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+
+			RuleBuilder b = new RuleBuilder();
+			myRule = (FhirQueryRuleImpl) b.allow()
+				.read()
+				.resourcesOfType("Patient")
+				.inCompartmentWithFilter("patient", myResource.getIdElement().withResourceType("Patient"), "family=Smith")
+				.andThen().build().get(0);
+
+			when(myMatcher.match(ArgumentMatchers.eq("Patient?family=Smith"), ArgumentMatchers.same(myResource)))
```
Comment (fil512) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImplTest.java:81]: this mock doesn't make sense to me.  Feels like this test would make more sense to just have a real Matcher.  Using a mock here makes this test kinda tautologous.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/AuthorizationInterceptor.java [hunk 12]
```diff
@@ -177,6 +179,16 @@ public AuthorizationInterceptor setValidationSupport(IValidationSupport theValid
 		return this;
 	}
 
+	/**
+	 * Sets a search parameter matcher for use in handling SMART v2 filter scopes
+	 *
+	 * @param theAuthorizationSearchParamMatcher The search parameter matcher. Null is also acceptable and is used as the
+	 *                                          default.   
+	 */
+	public void setMyAuthorizationSearchParamMatcher(IAuthorizationSearchParamMatcher theAuthorizationSearchParamMatcher) {
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/AuthorizationInterceptor.java:188]:
```suggestion
	public void setAuthorizationSearchParamMatcher(IAuthorizationSearchParamMatcher theAuthorizationSearchParamMatcher) {
```

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/AuthorizationInterceptor.java [hunk 13]
```diff
@@ -177,6 +179,16 @@ public AuthorizationInterceptor setValidationSupport(IValidationSupport theValid
 		return this;
 	}
 
+	/**
+	 * Sets a search parameter matcher for use in handling SMART v2 filter scopes
+	 *
+	 * @param theAuthorizationSearchParamMatcher The search parameter matcher. Null is also acceptable and is used as the
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/AuthorizationInterceptor.java:185]: Express this with the `@Nullable` annotation

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImpl.java [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+import org.apache.commons.lang3.builder.ToStringBuilder;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.instance.model.api.IIdType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import java.util.Set;
+
+// wipjv Ken - Should we complicate RuleImplOp instead?  Or enhance the testers?
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImpl.java:15]: subclassing feels like the right pattern.  I assume sometimes we instantiate the superclass and sometimes we instantiate the subclass so it's polymorphic

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImpl.java [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+	public void setFilter(String theFilter) {
+		myFilter = theFilter;
+	}
+
+	public String getFilter() {
+		return myFilter;
+	}
+
+	@Override
+	protected AuthorizationInterceptor.Verdict applyRuleLogic(RestOperationTypeEnum theOperation, RequestDetails theRequestDetails, IBaseResource theInputResource, IIdType theInputResourceId, IBaseResource theOutputResource, Set<AuthorizationFlagsEnum> theFlags, FhirContext theFhirContext, RuleTarget theRuleTarget, IRuleApplier theRuleApplier) {
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImpl.java:42]: Number of params here is a code smell, but again, not a problem we want to deal with today.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImpl.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+		ourLog.trace("applyRuleLogic {} {}", theOperation, theRuleTarget);
+		// wipjv hack alert - theOutputResource == null means we're in the pre-show pointcut.
+		if (theOutputResource == null) {
+			return super.applyRuleLogic(theOperation, theRequestDetails, theInputResource, theInputResourceId, theOutputResource, theFlags, theFhirContext, theRuleTarget, theRuleApplier);
+		}
+		IAuthorizationSearchParamMatcher matcher = theRuleApplier.getSearchParamMatcher();
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImpl.java:48]: I like how this method is put together.  Nice and clean.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/IAuthRuleBuilderRuleOpClassifier.java [hunk 17]
```diff
@@ -134,4 +133,8 @@ public interface IAuthRuleBuilderRuleOpClassifier {
 	 * @since 6.0.0
 	 */
 	IAuthRuleFinished withCodeNotInValueSet(@Nonnull String theSearchParameterName, @Nonnull String theValueSetUrl);
+
+	IAuthRuleFinished inCompartmentWithFilter(String theCompartment, IIdType theIdElement, String theFilter);
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/IAuthRuleBuilderRuleOpClassifier.java:137]: do these two need to be combined into one method?  Could we compose an inCompartment with a withFilter instead?

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/IAuthorizationSearchParamMatcher.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+public interface IAuthorizationSearchParamMatcher {
+	MatchResult match(String theCriteria, IBaseResource theResource);
+
+	enum Match {
+		MATCH,
+		NO_MATCH,
+		UNSUPPORTED
+	}
+
+	public static class MatchResult {
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/IAuthorizationSearchParamMatcher.java:19]: Why not just reuse InMemoryMatchResult ?

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/consent/RuleFilteringConsentService.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+	 * @param theResource        The resource that will be exposed
+	 * @param theContextServices Unused.
+	 * @return REJECT if the rules don't ALLOW, PROCEED otherwise.
+	 */
+	@Override
+	public ConsentOutcome canSeeResource(RequestDetails theRequestDetails, IBaseResource theResource, IConsentContextServices theContextServices) {
+		ourLog.trace("canSeeResource() {} {}", theRequestDetails, theResource);
+
+		// apply rules!  If yes, then yes!
+		// wipjv Ken - ask if this the right RestOperationTypeEnum?
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/consent/RuleFilteringConsentService.java:47]: I don't think I understand it well enough to answer.  Would need to read the code.  How is it usually determined?

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImplTest.java (hunk 11 above)
Comment (michaelabuckley) [reply hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImplTest.java:81]: A real matcher would require real search parameters and a whole DAO.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImpl.java (hunk 15 above)
Comment (michaelabuckley) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/FhirQueryRuleImpl.java:42]: Couldn't agree more.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/IAuthorizationSearchParamMatcher.java (hunk 18 above)
Comment (michaelabuckley) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/IAuthorizationSearchParamMatcher.java:19]: NOMATCH and UNSUPPORTED are different cases, and they are combined in the InMemoryMatchResult

**Pull request merged**